File: backend/src/main/java/vjda/resource/ProxyResource.java
package vjda.resource;

import java.net.URI;

import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.HttpMethod;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.HttpHeaders;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.UriInfo;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;

import vjda.proxy.ApiProxy;
import vjda.proxy.InvalidTargetException;
import vjda.resource.exception.BadRequestException;

@Path("proxy/{target}/{path:.*}")
public class ProxyResource {
	@Autowired private ApiProxy apiProxy;
	@Context UriInfo uriInfo;
	@Context HttpHeaders headers;

	@GET
	@Consumes(MediaType.APPLICATION_JSON_VALUE)
	@Produces(MediaType.APPLICATION_JSON_VALUE)
	public Response get(@PathParam("target") String target, @PathParam("path") String path) {
		return proxy(target, path, HttpMethod.GET, null);
	}

	@POST
	@Consumes(MediaType.APPLICATION_JSON_VALUE)
	@Produces(MediaType.APPLICATION_JSON_VALUE)
	public Response post(@PathParam("target") String target, @PathParam("path") String path, String entity) {
		return proxy(target, path, HttpMethod.GET, entity);
	}

	@PUT
	@Consumes(MediaType.APPLICATION_JSON_VALUE)
	@Produces(MediaType.APPLICATION_JSON_VALUE)
	public Response put(@PathParam("target") String target, @PathParam("path") String path, String entity) {
		return proxy(target, path, HttpMethod.GET, entity);
	}

	@DELETE
	@Produces(MediaType.APPLICATION_JSON_VALUE)
	public Response delete(@PathParam("target") String target, @PathParam("path") String path) {
		return proxy(target, path, HttpMethod.GET, null);
	}

	private Response proxy(String target, String path, String method, String entity) {
		URI uri = uriInfo.getRequestUriBuilder().replacePath(path).build();
		try {
			return apiProxy.proxy(target, uri, method, entity);
		} catch (InvalidTargetException e) {
			throw new BadRequestException("target is invalid.");
		}
	}
}

File: backend/src/main/kotlin/backend/resource/ProxyResource.kt
package backend.resource

import backend.proxy.ApiProxy
import backend.proxy.InvalidTargetException
import backend.resource.exception.BadRequestException
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import javax.ws.rs.*
import javax.ws.rs.core.Context
import javax.ws.rs.core.HttpHeaders
import javax.ws.rs.core.Response
import javax.ws.rs.core.UriInfo

@Path("proxy/{target}/{path:.*}")
class ProxyResource {
    @Autowired
    lateinit var apiProxy: ApiProxy

    @Context
    lateinit var uriInfo: UriInfo

    @Context
    lateinit var headers: HttpHeaders

    @GET
    @Consumes(MediaType.APPLICATION_JSON_VALUE)
    @Produces(MediaType.APPLICATION_JSON_VALUE)
    operator fun get(@PathParam("target") target: String, @PathParam("path") path: String): Response {
        return proxy(target, path, HttpMethod.GET, null)
    }

    @POST
    @Consumes(MediaType.APPLICATION_JSON_VALUE)
    @Produces(MediaType.APPLICATION_JSON_VALUE)
    fun post(@PathParam("target") target: String, @PathParam("path") path: String, entity: String?): Response {
        return proxy(target, path, HttpMethod.GET, entity)
    }

    @PUT
    @Consumes(MediaType.APPLICATION_JSON_VALUE)
    @Produces(MediaType.APPLICATION_JSON_VALUE)
    fun put(@PathParam("target") target: String, @PathParam("path") path: String, entity: String?): Response {
        return proxy(target, path, HttpMethod.GET, entity)
    }

    @DELETE
    @Produces(MediaType.APPLICATION_JSON_VALUE)
    fun delete(@PathParam("target") target: String, @PathParam("path") path: String): Response {
        return proxy(target, path, HttpMethod.GET, null)
    }

    private fun proxy(target: String, path: String, method: String, entity: String?): Response {
        val uri = uriInfo.requestUriBuilder.replacePath(path).build()
        return try {
            apiProxy.proxy(target, uri, method, entity)
        } catch (e: InvalidTargetException) {
            throw BadRequestException("target is invalid.")
        }
    }
}